core/templates/services/state-interaction-stats.service.ts

```diff
@@ -111,24 +108,26 @@ export class StateInteractionStatsService {
           exploration_id: explorationId,
           state_name: state.name,
         }))
-      .toPromise().then(response => <IStateRulesStats>{
-        exploration_id: explorationId,
-        state_name: state.name,
-        visualizations_info: response.visualizations_info.map(info => ({
-          addressed_info_is_supported: info.addressed_info_is_supported,
-          data: info.data.map(datum => <IAnswerData>{
-            answer: this.getReadableAnswerString(state, datum.answer),
-            frequency: datum.frequency,
-            is_addressed: (
-              info.addressed_info_is_supported ?
+      .toPromise().then(response => {
+        this.cachedStats = {
```

Comment:
Instead of caching the result, cache the promise, so that when `computeStats` is called twice in a short period of time (so the first request is not yet completed) the second HTTP is not made and it just waits for the firs one too.

Reply:
Hang on -- but in the long run, the promise would be resolved. So what happens if the second call comes after a long period of time?

Reply:
The promise is resolved but still exists as an object so the data can be retrieved from it without calling the endpoint again.

Reply:
Oh, interesting, I see, thanks.

Sorry @brianrodri -- I didn't realize this when I commented. Could you please add a comment above the cached promise to explain why we're doing it that way? (Also @vojtechjelinek do you have a link that we can put here as a reference to explain this pattern?)

Thanks!

Reply:
Not really

Reply:
Resolved? @seanlip